File: lab-2/Model/PlayerModule/PlayerSet.js
/**
 * a class that represents the profile of one or more players participating in this game
 * 
 * @member {Array} playersList - Array of players
 * @member {number} numOfPlayers - number of players participating in this game
 * 
 * @author Songyuan Wu & Jing Wen
 */

class PlayerSet {
	/**
	 * initialize the players list
	 *
	 * @returns {PlayerSet} a new instance of type Players
	 */
	constructor() {
		this.playersList = [];
		this.numOfPlayers = 0;
	}

	/**
	 * add a computer player to the game.
	 *
	 * @require the the player is not the game
	 * @param {number} points points that the player won or lost.
	 * @returns {ComputerPlayer} the player that just added.
	 * @returns {null} if the name is aready exsit.
	 */
	 addComputerPlayer(name, winProb) {
		if (this.playerExist(name)) {
			return null;
		}
		this.numOfPlayers += 1;
		var player = new ComputerPlayer(name);
		player.winProb= winProb;
		this.playersList.push(player);

		return player;
	}

	/**
	 * add a human player to the game.
	 *
	 * @require the the player is not the game
	 * @param {string} name the name of the specific player
	 * @returns {Player} the player that just added.
	 * @returns {null} if the name is aready exsit.
	 */
	 addPlayer(name) {
		if (this.playerExist(name)) {
			return null;
		}
		this.numOfPlayers += 1;
		var player = new Player(name);
		this.playersList.push(player);

		return player;
	}

	/**
	 * delete a player to the game.
	 *
	 * @require the the player is not the game
	 * @param {string} name the name of the specific player
	 * @returns {Player} the player that just deleted
	 * @returns {null} if player is not in the game
	 */
	 deletePlayer(name) {
		for (var element of this.playersList) {
			if (element.playerName== name) {
				this.numOfPlayers -= 1;
				
				var idx=this.playersList.map(x=>x.playerName).indexOf(element.playerName)
				this.playersList.splice(idx,1)
				// copy_list = this.playersList;
				// this.playersList = [];
				// for (let i = 0; i < copy_list; i++) {
				//	 if (copy_list[i] != name) {
				//		 this.playersList.push(copy_list[i]);
				//	 }
				// }
				return element;
			}
		}
		return null;
	}

	/**
	 * return a specific player's score.
	 *
	 * @require the the player is in the game
	 * @param {string} name the name of the specific player
	 * @returns {Player} that player you are looking for
	 * @returns {null} if the player does not exsit.
	 */
	 playersSearch(name) {
		for (var element of this.playersList) {
			if (element.playerName== name) {
				return element;
			}
		}
		return null;
	}

	/**
	 * reset all player's score.
	 */
	 resetAllScores() {
		for (var element of this.playersList) {
			element.scoreChangeTo(0);
		}
	}

	/**
	 * return a specific player's score.
	 *
	 * @param {string} name the name of the specific player
	 * @returns {number} the score of one specific player
	 * @returns {null} if player does not exist.
	 */
	 playersScoreCheck(name) {
		for (var element of this.playersList) {
			if (element.playerName== name) {
				return element.playerScore;
			}
		}
		return null;
	}

	/**
	 * display the highest score in this game
	 *
	 * @returns {number} the highest score
	 */
	 playersHighestScore() {
		var highest_score = 0;
		for (var element of this.playersList) {
			if (element.playerScore > highest_score) {
				highest_score = element.playerScore;
			}
		}
		return highest_score;
	}

	/**
	 * check is the player already in the game
	 *
	 * @param {string} name the name of a specific player
	 * @returns {boolean} true the player is already in the game
	 */
	 playerExist(name) {
		var result = false;
		for (var element of this.playersList) {
			console.log(element)
			if (element.playerName == name) {
				result = true;
				return result;
			}
		}
		return result;
	}

	/**
	 * add the socre to a player
	 *
	 * @param {string} name the name of a specific player
	 * @param {int} score the score to add
	 * @returns {boolean} true the player is already in the game
	 */
	 addScore(name,score) {
		var result = false;
		for (var element of this.playersList) {
			console.log(element)
			if (element.playerName == name) {
				result = true;
				element.scoreAdd(score);
				return result;
			}
		}
		return result;
	}


	/**
	 * check is the player already in the game
	 *
	 * @param {string} name the name of a specific player
	 * @returns {boolean} true the player is already in the game
	 */
	 playersName() {
		players = [];
		for (var element of this.playersList) {
			players.concat(element.playerName);
		}
		return players;
	}
}
	

File: lab-1/PlayersModule/player-set.rb
class Players
  # a class that represents the profile of one or more players participating in this game.
  #
  # @attr_reader no_of_players [Number] number of players participating in this game
  # @attr_reader players_list [Array Object] storing all of the players' information

  @@no_of_players = 0
  @@players_list = Array.new

  attr_reader :no_of_players
  attr_reader :players_list
  # initialize each player's profile
  #
  # @param name [Name, #read] the name of the specific player
  # @return an Array object which contains the new player's information
  def player_initialize(name)
    @@no_of_players += 1

    player_info = Hash["player_name" => name, "player_score" => 0]
    @@players_list.append(player_info)
  end

  # update each player's profile
  #
  # @param name [Name, #read] the name of the specific player
  # @param points [Number, #read] the points of the specific player who win or loss, as well as neither
  # @return an Array object which contains the players' updated information
  def player_property_update(name, points)
    for element in @@players_list
      if element["player_name"] == name
        element["player_score"] += points

        element["player_score"] = 0 if element["player_score"] < 0
      end
    end
  end

  # display each player's profile
  #
  # @return an Array object which contains all of players' information
  def all_players_status
    i = 1
    for element in @@players_list
      puts "Player #{i} #############"
      puts "Player Name: #{element["player_name"]}"
      puts "Player Score: #{element["player_score"]}"
      i += 1
    end
  end

  # display one specific player's profile
  #
  # @return an Array object which contains the specific player's information
  def specific_player_status(name)
    for element in @@players_list
      if element["player_name"]  == name
        puts "Player Name: #{element["player_name"]}"
        puts "Player Score: #{element["player_score"]}"
      end
    end
  end

  # report a specific player's score
  #
  # @param name [Name, #read] the name of a specific player
  # @return the score of one specific player
  def player_score(name)
    for element in @@players_list
      if element["player_name"]  == name
        @player_score = element["player_score"]
      end
    end
    return @player_score
  end

  # report the total number of players
  #
  # @return the total number of players
  def total_no_of_players
    puts "Total number of players: #@@no_of_players"
  end

end